### PR TITLE
[Feature][Connector-V2] Support OpenMLDB multi-table source

### PR DESCRIPTION
**Pull request by goutamadwant**
### Purpose of this pull request

Related to #10425, covering the OpenMldb source only.

Add multi-table reads using the common `tables_configs` option. Each query declares its result schema and output table identity and can override the source database. Connection settings remain shared.

The reader changes also address correctness problems encountered while validating this path: SQL NULL handling, nullable metadata, reused row arrays, shared executor ownership, and cluster queries taking the offline-job execution path.

### Does this PR introduce _any_ user-facing change?

Yes.

Before, one source accepted one root-level `sql` query and used the SDK's input-schema discovery. Separate queries with different result schemas required separate source configurations.

After, one source can read multiple queries sequentially through `tables_configs`:

- Each entry requires `sql`, a unique `schema.table`, and non-empty `schema.fields`; `database` is optional per entry.
- Result columns are matched to configured field names, including SQL aliases, case-sensitively. Incorrect names, counts, or types fail the read.
- Rows carry their output table identity. Batch completion is reported only after all queries succeed, including empty results.
- SQL NULL values remain null for all supported types, and closing one reader does not close another reader's connection.
- Cluster reads query online table rows directly instead of submitting an offline job.

Existing root-level `sql` configurations remain supported. Their input-schema and positional-mapping limitations are documented. No connector options are removed, connection defaults are unchanged, and the SDK remains at 0.6.3.

This does not add parallelism, CDC, exactly-once delivery, or a consistent snapshot across tables. Streaming mode continues to repeat the configured queries and can emit duplicates.

English and Chinese source documentation include configuration examples and validation rules.

### How was this patch tested?

- 29 connector unit tests passed on both Java 8 and Java 11.
- The Java 11 connector dependency reactor passed `verify`: 506 tests, no failures, errors, or skipped tests.

```shell
./mvnw -o -pl seatunnel-connectors-v2/connector-openmldb -am verify
```

- Five added integration tests passed against OpenMLDB 0.6.3 in each combination of Java 8/11 and standalone/cluster mode: 20 successful test executions. These were run through JUnit Platform in a compatible Linux amd64 container against real servers, separately from Maven verification.
- Coverage includes different databases and schemas, reordered aliases, all supported nullable types, zero/false/empty-string values, empty results, reader isolation, query failures, and completion behavior.
- The integration tests also verify 1,101 complete, unique rows, using four partitions in cluster mode.

`OpenMldbSourceIT` is opt-in with `-Dopenmldb.integration=true`. Its class documentation lists the standalone and ZooKeeper connection properties. It requires a disposable server and a Linux amd64 runtime compatible with the pinned SDK's native library.

Zeta/Flink/Spark engine E2E validation remains pending. The real-server connector tests above do not replace that matrix; this change should remain draft while those checks are completed.

### Check list

- [x] License notice reviewed: no new dependencies or JAR binaries.
- [x] English and Chinese connector documentation updated.
- [x] Compatibility reviewed: no configuration migration required; reader behavior corrections are described above.
- [x] Existing connector registration reviewed: plugin mapping, distribution configuration, CI labels, and plugin configuration require no changes.
- [x] Unit tests and opt-in real-server connector integration tests added.
- [ ] Engine E2E testcase and Zeta/Flink/Spark validation.

### CODE REVIEW

**Review by DanielLeens — COMMENTED** [review]
# What Problem Does This PR Solve?

The existing `connector-openmldb` source can only read a single query result as a single table. This PR adds a `tables_configs`-based multi-table mode (following the same `ConnectorCommonOptions.TABLE_CONFIGS` pattern used by other recently-added multi-table sources in this repo, e.g. the just-merged IoTDB multi-table source) so one `OpenMldb` source can read several independently-schemed queries — even across different databases on the same cluster — into separate downstream tables in one job.

One-sentence summary: this is a genuine multi-table feature, but it also rewrites the connector's native-SDK read path and quietly fixes two real, pre-existing correctness bugs along the way (NULL values silently coerced to zero/false, and an inverted nullable-flag), neither of which is called out as a "fix" in the PR title/description even though they change output for the *existing* single-table users too.

# 1. Code Change Review

## 1.1 Core Logic Analysis

**Multi-table wiring.** `OpenMldbSourceFactory.optionRule()` uses `.exclusive(SQL, TABLE_CONFIGS)`. I checked `ConfigValidator.checkExclusive()` (`seatunnel-api/.../ConfigValidator.java:437-460`) directly rather than assuming: it throws if **zero** of the two are set ("exactly one option must be set, but none are configured") and if more than one is set — so `createSource()`'s fallback to the single-`sql` path can never silently run with no query configured. Good defensive validation, not just documentation.

`OpenMldbSourceFactory.TablesValidator` (a `ConditionExtension`) is thorough: it rejects any table-config key outside `sql`/`database`/`schema` (connection options must live at source level, matching the docs), requires a non-empty `schema.fields`, walks every declared field type against a hard-coded allow-list of supported `SqlType`s, and rejects duplicate table identities via a `Set<String> ids`. This validation happens at config-build time, before any connection is opened.

**Runtime dispatch (`OpenMldbSourceReader.pollNext`)**: one shared `OpenMldbReadClient` (opened once in `open()` against `tableParameters.get(0)`'s connection settings) is reused for every table's query in a sequential `for` loop, passing each table's own `database` string into `client.execute(database, sql, rowType, tableId != null)`. I verified this is safe specifically because `OpenMldbSourceFactory.buildTables()` builds every table's config via `ReadonlyConfig.fromMap(entry).toConfig().withFallback(sourceLevelConfig...)` — i.e. every table entry inherits the *same* `host`/`port`/`zk_host`/`cluster_mode`/timeouts from the source level (`TablesValidator` doesn't even allow overriding them), so one shared native `SQLRouter` really can serve every table; only `database` (and optionally `sql`) vary per call, and `execute()` takes `database` as a per-call argument rather than assuming it from construction. This is a correct design, not an accidental shortcut.

**Native resource handling in the new `OpenMldbReadClient`** (a wholly new file wrapping the native `SQLRouter`/`ResultSet`/`Schema`/`Status`/`Date` JNI objects): I traced the ownership carefully because this is exactly the kind of code that leaks native memory or double-frees if done casually.
- `execute()` only nulls its local `result` reference *after* the `Query` constructor succeeds; if `Query`'s constructor throws (e.g. column-count/type mismatch), the `finally` block still deletes the half-built `ResultSet`. Correct.
- `Query`'s constructor deletes the `Schema` handle in a `finally` regardless of validation outcome.
- `readValue()` for `DATE` deletes the native `Date` struct in a `finally` after converting to `LocalDate`.
- `execute()` and `close()` are both `synchronized` on the client instance — this matters because a checkpoint/cancellation thread calling `close()` concurrently with an in-flight `pollNext()` on the native `router` is a real use-after-free risk for JNI-backed objects; synchronizing both paths on the same monitor closes that race properly rather than leaving it to luck.

**Two bug fixes bundled into this "Feature" PR** (verified by diffing against the pre-PR `dev` version of these files, not assumed from the description):
1. The pre-PR reader used plain JDBC (`java.sql.ResultSet` via `sqlExecutor.executeSQL(...)`) and called `resultSet.getInt/getBoolean/...(index)` with **no `wasNull()` check at all** — JDBC primitive getters return `0`/`false` for SQL `NULL`, so every `NULL` in a nullable numeric/boolean column was silently corrupted to a zero-like value. The new `OpenMldbReadClient.Query.readRow()` explicitly checks `result.IsNULL(columnIndexes[i])` first and leaves the field `null` if so — this is a real, valuable data-correctness fix, not a refactor.
2. `PhysicalColumn.of(name, type, columnLength, boolean nullable, ...)` (confirmed the parameter is `nullable`, `seatunnel-api/.../PhysicalColumn.java:213-221`) was previously called with `column.isNotNull()` passed directly as the `nullable` argument — inverted. This PR changes it to `!column.isNotNull()`, which is correct.

Both fixes apply to the **existing single-table code path**, not just the new multi-table one, so this is a behavior change for current OpenMldb-source users: a column that used to silently read as `0`/`false`/empty for `NULL` will now correctly read as `null` after upgrade, and the produced `CatalogTable`'s nullable metadata will flip for any not-null/nullable column. That's the right fix, but it's real user-facing behavior change riding on a "Feature" PR title, and it isn't mentioned in the PR description at all (see Issue 1 below) — it *is*, to the author's credit, disclosed prominently in `docs/en/connectors/source/OpenMldb.md` ("SQL `NULL` values remain `null` ... they are not converted to zero or `false`"), so it's not hidden from users reading the docs, just missing from the PR's own changelog-facing description.

## 1.2 Compatibility Impact

**Partially incompatible, in the good-bug-fix sense.** No config option is renamed or removed, and `sql`-mode (single-table) config remains valid and unchanged in shape. But behavior for existing single-table jobs does change on upgrade for any table with nullable numeric/boolean/string columns actually containing `NULL` at read time — values that previously arrived as `0`/`false`/`""` will now correctly arrive as `null`. Any downstream transform/sink that was implicitly relying on the old (wrong) zero-coercion behavior would see different output after upgrading. Given the old behavior was a genuine correctness bug rather than a documented contract, I don't think this should block merge, but per this repo's own compatibility policy it should be called out explicitly as a behavior-affecting fix in the PR description (it already is in the docs).

## 1.3 Performance / Side-Effect Analysis

Multi-table reads are strictly sequential within one reader instance — `OpenMldbSource` extends `AbstractSingleSplitSource`, so there is exactly one split/one task regardless of configured `parallelism`, same as the pre-existing single-table connector and the same base class used by ~20 other non-partitioned connectors in this repo (Redis, Neo4j, the other HTTP connectors, etc.), so this isn't a new limitation introduced here — it's inherited from the framework choice, and it's honestly disclosed in the docs ("Multi-table reading does not add parallelism"). One shared native `SQLRouter` for all tables is the right call resource-wise (avoids N native connections for N tables). The streaming-mode caveat — `pollNext()` re-runs every configured query from scratch on each poll with no incremental cursor, so `job.mode = STREAMING` against this source can produce duplicate rows — is pre-existing (unchanged `getBoundedness()`/loop-termination logic) and is now explicitly documented for both single- and multi-table mode ("this is not CDC or incremental polling and can produce duplicate records"). I'd call this an acceptable, disclosed limitation rather than a blocking defect.

## 1.4 Error Handling and Logging

`readTable()` wraps any `SQLException`/`RuntimeException` from a single table's query into an `OpenMldbConnectorException` that names the failing `tableId` (or database, for single-table mode) — a failure on table 2 of 5 is clearly attributable rather than an opaque stack trace. `Query`'s constructor fails fast with a specific message for column-count mismatch, missing-by-name column, and type mismatch, rather than silently truncating or coercing.

# 2. Code Quality Assessment

## 2.1 Coding Standards

New/changed non-trivial methods and fields have explanatory comments describing intent (e.g. `OpenMldbReadClient`'s class comment "Owns the native resources needed to read nullable query results with SDK 0.6.3", the "Ownership transfers only after schema validation succeeds" comment on the `result = null` line, "JDBC primitive getters return zero/false for NULL, and wasNull is unsupported" on the null-check). This is exactly the kind of "explain why, not what" comment that documents a genuinely non-obvious constraint, not boilerplate.

`OpenMldbSqlExecutor` used to hold a **process-wide static mutable singleton** (`private static volatile SqlClusterExecutor SQL_EXECUTOR` plus a shared mutable `SDK_OPTION`), with a static `close()` that would tear down the executor for every other `OpenMldbSource` instance in the same JVM. This PR replaces it with per-call `create()`/`createReader()` factory methods with no shared mutable state — a real fix for a static-state hazard that would have caused two differently-configured OpenMldb sources (or two subtasks) in the same JVM to silently interfere with each other. I grepped for remaining references to the removed `initSdkOption()`/`getSqlExecutor()`/static `close()` and found none — no dangling callers.

## 2.2 Test Coverage and Test Stability

`OpenMldbFactoryTest` covers the `optionRule`/`TablesValidator` edge cases (blank table id, unsupported keys, duplicate identities, unsupported types). `OpenMldbSourceTest` and `OpenMldbSourceReaderTest` are pure Mockito-based unit tests using `mockConstruction(OpenMldbReadClient.class, ...)` — no real network or native SDK involved. `OpenMldbSourceReaderTest` specifically verifies: rows from different tables carry the correct `tableId`; an empty first table's `Query` doesn't cause the loop to skip the second table; `pollNext()` called twice only signals `signalNoMoreElement()` once (i.e., the `finished` guard actually prevents a second full re-read in bounded mode); and every `Query`/client resource is verified `.close()`d exactly once even with `reader.close()` called twice (idempotency). This is genuinely edge-case-directed testing, not just happy-path.

`OpenMldbSourceIT` is a real-server integration test but is gated behind `@EnabledIfSystemProperty(named = "openmldb.integration", matches = "true")` — it does **not** run in default CI (the class Javadoc is explicit that it needs a disposable OpenMLDB 0.6.3 server and a Linux amd64 native runtime). This is an honest, non-flaky way to gate a native-SDK-dependent test, consistent with how other native-binary-constrained tests in this repo are handled, but it does mean the multi-table path's true behavior against a live OpenMLDB server (real NULL semantics from the native layer, real cross-database queries) is verified by a human/CI-with-the-flag-set, not by default `mvn test`. The mock-based tests compensate well for the dispatch/lifecycle logic but can't validate the native JNI layer itself.

Flaky-test rating: **Stable**. All default-CI-executed tests are deterministic pure-unit/mock tests with no timing, network, or shared static state (the removal of the old static singleton actually *reduces* future flake risk here, since parallel test classes can no longer stomp on each other's `SDK_OPTION`).

## 2.3 Documentation Updates

Both `docs/en/connectors/source/OpenMldb.md` and `docs/zh/connectors/source/OpenMldb.md` are updated and consistent with each other (I diffed both) and updated to match the code exactly: the `tables_configs` option, its `sql`/`database`/`schema` sub-fields, the NULL-preservation behavior, the streaming-mode duplicate-read caveat, the "no parallelism / no cross-table snapshot / no exactly-once" caveats, and a full multi-table example config. This is unusually thorough and honest documentation — every non-obvious limitation I independently derived from reading the code was already called out here before I finished my own analysis.

# 3. Architectural Soundness

## 3.1 Elegance of the Solution

Long-term solution, not a workaround. Reusing `ConnectorCommonOptions.TABLE_CONFIGS` (the same convention as the recently-merged IoTDB multi-table source) rather than inventing a bespoke multi-table config shape keeps this consistent with the rest of the codebase. Rewriting the native-SDK read layer to fix the NULL-handling and nullable-flag bugs as a prerequisite for multi-table support (rather than layering multi-table on top of the broken JDBC-getter path) was the right call, even though it does mean this PR is doing two things at once (see Issue 1).

## 3.2 Maintainability

Good — the native resource ownership is carefully commented and the client/query split (`OpenMldbReadClient` vs. `OpenMldbSourceReader`) keeps the sequential per-table dispatch logic separate from the native-SDK plumbing.

## 3.3 Extensibility

The `TablesValidator` pattern and the shared-client design leave room for adding more per-table overrides later if ever needed (though currently only `sql`/`database`/`schema` are allowed per table, which is a deliberate, documented restriction, not an oversight).

## 3.4 Historical-Version Compatibility

No stored/checkpoint/catalog format is touched — this is a stateless batch (or, in existing streaming usage, re-polling) source. The behavior change is confined to in-flight value semantics (NULL handling, nullable metadata), not to any persisted format, so there's no restore/upgrade-path risk, only the output-semantics change discussed in 1.2.

# 4. Issue Summary

| # | Issue | Location | Severity |
|---|-------|----------|----------|
| 1 | The PR fixes two real pre-existing bugs affecting the *existing* single-table path (NULL-to-zero/false coercion; inverted `nullable` flag on `PhysicalColumn`) but the PR title/description only describes this as adding multi-table support — the behavior change isn't mentioned in the PR description (it is disclosed in `docs/en`+`docs/zh`, just not in the description/changelog-facing text). | PR description; `seatunnel-connectors-v2/connector-openmldb/src/main/java/.../source/OpenMldbSource.java` (nullable-flag fix), `OpenMldbReadClient.java` (NULL-handling fix) | Medium |
| 2 | `OpenMldbSourceIT`, the only test that exercises the new client against a real OpenMLDB server, is skipped by default (`@EnabledIfSystemProperty`) and never runs in standard CI. The mock-based unit tests cover dispatch/lifecycle logic well but can't validate the native JNI layer's actual NULL/type behavior end-to-end. | `seatunnel-connectors-v2/connector-openmldb/src/test/java/.../source/OpenMldbSourceIT.java` | Low |

# 5. Merge Recommendation

### Conclusion: Ready to merge

1. Blockers — must be fixed: none found in the source. The one open item before this can actually be merged is procedural, not code: **CI on the fork is still queued** at the time of this review (only `License header`/`changes`/`Code style`/`Dead links` had even started, all `queued`, nothing completed yet) — I'm not able to confirm a green build yet, and per this repo's policy I won't claim "verified" CI. Please make sure the fork's full run (including `unit-test` and any `updated-modules-integration-test-*` jobs this module maps to) completes and is green before merge.
2. Recommended fixes — non-blocking:
   - Issue 1: add a line to the PR description calling out the NULL-handling and nullable-flag fixes as behavior changes for existing single-table jobs (the docs already say this well; the description should too, since that's what typically feeds release notes).
   - Issue 2: consider, as a future follow-up (not this PR), whether a Testcontainers-based OpenMLDB image could let at least a smoke-level multi-table test run in default CI — understanding this may not be practical given the native SDK's platform constraints.

Overall assessment: this is a well-engineered PR. The multi-table wiring is validated defensively at config time, the native resource management in the new client is careful about ownership and concurrent-close safety (a real concern for JNI-backed objects under task cancellation), two genuine pre-existing correctness bugs are fixed as a side effect, and the documentation is unusually thorough and honest about every limitation (streaming duplicates, no parallelism, no cross-table snapshot) — points I'd normally have to raise myself were already disclosed. I don't see a better alternative architecture here; sequential single-split reads are the correct trade-off for this connector's native-SDK constraints, and the shared-router-with-per-call-database design is the right way to avoid N native connections for N tables. Once CI is confirmed green and the PR description is updated to mention the NULL/nullable-flag fix, this is good to merge.

File: seatunnel-connectors-v2/connector-openmldb/src/main/java/org/apache/seatunnel/connectors/seatunnel/openmldb/source/OpenMldbReadClient.java [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+            throws SQLException {
+        if (router == null) {
+            throw new SQLException("OpenMldb reader is closed");
+        }
+        Status status = new Status();
+        ResultSet result = null;
+        try {
+            // Query the online tables directly; ExecuteSQL can submit an offline job in cluster
+            // mode.
+            result = router.ExecuteSQLParameterized(database, sql, null, status);
```
Comment (goutamadwant) [inline seatunnel-connectors-v2/connector-openmldb/src/main/java/org/apache/seatunnel/connectors/seatunnel/openmldb/source/OpenMldbReadClient.java]: `ExecuteSQLParameterized` uses OpenMLDB's batch-query path. In the pinned 0.6.3 server, `TabletImpl::ProcessQuery` truncates output after `scan_max_bytes_size`—2 MiB by default—but returns `kOk`. This code therefore accepts a partial result, and the bounded reader subsequently signals successful end-of-input, silently losing rows for ordinary query results above 2 MiB. The 1,101-row integration case remains below that byte limit. Please use a path that can exhaust or paginate the result, or otherwise prevent a truncated response from being reported as complete.

**Comment by goutamadwant** [comment]
@DanielLeens Update in 1e436ea42:

Renamed the opt-in integration test to `TestOpenMldbSourceIT` to satisfy the repository naming rule, without changing its gate or behavior. The naming check and 29 module tests pass; the live integration test remains intentionally skipped. Fresh CI has been triggered. The opt-in reproduction command should now use the new class name; live database and engine E2E verification remain separate.

**Review by DanielLeens — COMMENTED** [review]
# What Problem Does This PR Solve?

The existing `connector-openmldb` source could only read a single query result as a single table. This PR adds a `tables_configs`-based multi-table mode (the same `ConnectorCommonOptions.TABLE_CONFIGS` pattern used by the recently-merged IoTDB multi-table source) so one `OpenMldb` source can read several independently-schemed queries, even across different databases on the same cluster, into separate downstream tables in one job. Along the way it rewrites the connector's native-SDK read path and fixes several real, pre-existing correctness bugs (NULL values silently coerced to zero/false, an inverted nullable flag, reused row arrays, a shared mutable static executor, and offline-job execution for cluster reads) that also affect the *existing* single-table users. One-sentence summary: this is a genuine multi-table feature that also quietly repairs a handful of real bugs in the connector it's built on top of, and both are now clearly disclosed in the PR description and docs.

This is a full re-review of the current head, not an incremental diff check, since a new commit landed after my last review.

# 1. Code Change Review

## 1.1 Core Logic Analysis

I re-verified the core claims from my previous review directly against the current head rather than trusting the earlier pass:

- `OpenMldbSourceFactory.optionRule()` still uses `.exclusive(SQL, TABLE_CONFIGS)` (`OpenMldbSourceFactory.java:62`). `ConfigValidator.checkExclusive()` throws if zero or more than one of the two are set, so `createSource()` cannot silently fall back to an unconfigured single-`sql` path.
- `TablesValidator` still rejects unknown per-table keys, requires non-empty `schema.fields`, checks every field type against an allow-list, and rejects duplicate table ids.
- `OpenMldbSourceReader.pollNext()` still reuses one shared `OpenMldbReadClient` across every configured table, passing each table's own `database` string per call; this is safe because `TablesValidator` doesn't allow overriding connection-level settings per table, so one native `SQLRouter` genuinely can serve every table.
- In the new `OpenMldbReadClient.java`, I re-confirmed `execute()` (line 48) and `close()` (line 75) are both `synchronized` on the client instance, closing the concurrent-close-vs-in-flight-poll race that would otherwise be a real use-after-free risk for the JNI-backed `SQLRouter`/`ResultSet`. `readRow()` (around line 132) checks `result.IsNULL(columnIndexes[i])` before reading, with the comment "JDBC primitive getters return zero/false for NULL, and wasNull is unsupported" still present verbatim, and allocates a fresh `Object[] fields` per row, i.e. the row-array-reuse bug the PR description also calls out is genuinely fixed in this new client. `OpenMldbSource.java:153` still passes `!column.isNotNull()` into `PhysicalColumn.of(...)`, correcting the previously-inverted nullable flag.

**What actually changed since my last review:** exactly one file, a pure rename —
`seatunnel-connectors-v2/.../source/OpenMldbSourceIT.java` -> `TestOpenMldbSourceIT.java`, with the two in-file self-references (`OpenMldbSourceIT.class` -> `TestOpenMldbSourceIT.class`) updated to match. I diffed the two commits directly (`git diff 4666afc4 1e436ea4`) and confirmed there is no other content change anywhere in the diff.

**Why this rename was necessary (root-cause verified, not assumed):** the prior fork CI run (`goutamadwant/seatunnel` run `34788427810`) failed exactly one job, `Run / Code style`, with:
```
[ERROR] UTClassNameCheckTest.checkUTClassName:106 UT class does not conform to the naming convention,
must should be start with 'Test' or end with 'Test' or end with 'Tests' or end with 'TestCase'.
 .../connector-openmldb/src/test/java/.../source/OpenMldbSourceIT.java ==> expected: <0> but was: <1>
```
`OpenMldbSourceIT` matched none of the required patterns (it doesn't start with `Test` or end with `Test`/`Tests`/`TestCase`); `TestOpenMldbSourceIT` starts with `Test` and satisfies the check. This is a real, previously-failing CI gate, and the fix is the minimal, correct one.

**I also checked whether the rename changes runtime test-execution behavior, not just the naming-lint result.** The root `pom.xml` surefire config excludes `**/*IT.java` from the default `test` phase (`<exclude>**/*IT.java</exclude>`). `TestOpenMldbSourceIT.java` still ends in `IT.java`, so it is still excluded from `mvn test` under the new name — the rename fixes the naming-convention lint without changing which tests actually execute in default CI. The class also keeps `@EnabledIfSystemProperty(named = "openmldb.integration", matches = "true")`, so even if it were picked up, it would report as skipped rather than attempt a live connection. This exact `Test*IT` naming pattern also has precedent elsewhere in the repo (e.g. `TestVitessSourceReaderIT`, the `seatunnel-e2e/seatunnel-transforms-v2-e2e` `Test*IT` classes), so this isn't a one-off invention.

## 1.2 Compatibility Impact

Unchanged from my last review: **Partially incompatible, in the good-bug-fix sense.** No config option is renamed or removed; `sql`-mode remains valid. Behavior changes on upgrade for existing single-table jobs with nullable numeric/boolean/string columns actually containing `NULL` — those values now correctly arrive as `null` instead of `0`/`false`/`""`. This is a genuine correctness fix, not a documented contract break, so it shouldn't block merge, but it is user-facing.

I checked the current PR description specifically to see whether my previous non-blocking request ("call this out in the description, not just the docs") was addressed: it now reads, "The reader changes also address correctness problems encountered while validating this path: SQL NULL handling, nullable metadata, reused row arrays, shared executor ownership, and cluster queries taking the offline-job execution path," and the "Does this PR introduce any user-facing change?" section spells out the NULL-preservation behavior explicitly. **This resolves my prior recommendation.**

## 1.3 Performance / Side-Effect Analysis

Unchanged: multi-table reads are strictly sequential within one reader instance (`OpenMldbSource` extends `AbstractSingleSplitSource`, same base class as ~20 other non-partitioned connectors in this repo), so there's exactly one split/task regardless of configured parallelism — inherited from the framework choice and honestly documented, not a new limitation from this PR. One shared native `SQLRouter` for all tables avoids N native connections for N tables, which is the right resource trade-off. The streaming-mode re-poll-from-scratch caveat (possible duplicate rows under `job.mode = STREAMING`) is pre-existing and now explicitly documented for both modes.

## 1.4 Error Handling and Logging

Unchanged: `readTable()` wraps `SQLException`/`RuntimeException` per-table into an `OpenMldbConnectorException` naming the failing `tableId`/database, so a failure on table N of M is directly attributable. `Query`'s constructor fails fast with specific messages for column-count mismatch, missing-by-name column, and type mismatch.

# 2. Code Quality Assessment

## 2.1 Coding Standards

New/changed non-trivial methods and fields still carry explanatory "why" comments (native-resource-ownership comments in `OpenMldbReadClient`, the NULL-handling comment). The old process-wide static mutable `SqlClusterExecutor`/`SDK_OPTION` singleton in `OpenMldbSqlExecutor` is still replaced by per-call `create()`/`createReader()` with no shared mutable state; I re-grepped for the removed `initSdkOption()`/`getSqlExecutor()`/static `close()` and found no dangling callers.

## 2.2 Test Coverage and Test Stability

`OpenMldbFactoryTest`, `OpenMldbSourceTest`, and `OpenMldbSourceReaderTest` are unchanged, pure Mockito-based unit tests (`mockConstruction(OpenMldbReadClient.class, ...)`), still covering: rows from different tables carry the correct `tableId`; an empty first table doesn't cause the loop to skip the second; `pollNext()` called twice signals `signalNoMoreElement()` only once; every native resource is closed exactly once even under a double `reader.close()`. `TestOpenMldbSourceIT` (the renamed file) is unchanged in substance and remains gated behind `@EnabledIfSystemProperty(named = "openmldb.integration", matches = "true")`, i.e. it still does not run in default CI (confirmed via the surefire-exclude check in 1.1, not just the annotation).

Flaky-test anti-pattern check: no `Thread.sleep`, no nondeterministic external state, no shared static state (the removed singleton actually reduces future flake risk), no timing races in the mock-based tests. **Stability rating: Stable.**

## 2.3 Documentation Updates

`docs/en/connectors/source/OpenMldb.md` and `docs/zh/connectors/source/OpenMldb.md` are unchanged since my last pass and still consistent with each other and with the code: `tables_configs`, its sub-fields, NULL-preservation, the streaming-duplicate caveat, and the no-parallelism/no-cross-table-snapshot caveats are all documented with a full multi-table example.

# 3. Architectural Soundness

## 3.1 Elegance of the Solution (Long-term solution)

Reusing `ConnectorCommonOptions.TABLE_CONFIGS` rather than a bespoke shape keeps this consistent with the rest of the codebase. Fixing the NULL-handling/nullable-flag/row-reuse bugs as a prerequisite for multi-table support, rather than layering multi-table on the broken JDBC-getter path, was the right call.

## 3.2 Maintainability

Good — native resource ownership is carefully commented, and the client/reader split keeps native-SDK plumbing separate from per-table dispatch logic.

## 3.3 Extensibility

`TablesValidator` and the shared-client design leave room for future per-table overrides if ever needed; currently restricting overridable keys to `sql`/`database`/`schema` is a deliberate, documented restriction.

## 3.4 Historical-Version Compatibility

No stored/checkpoint/catalog format is touched; this is a stateless batch (or re-polling streaming) source. The only behavior change is in-flight value semantics (NULL handling, nullable metadata), not any persisted format.

# 4. Issue Summary

| # | Issue | Location | Severity |
|---|-------|----------|----------|
| 1 | *(Resolved since last review)* Previously: the PR fixed pre-existing bugs affecting the existing single-table path but didn't call this out in the PR description. The current PR description now explicitly lists these as correctness fixes and states the user-facing NULL-preservation change. No action needed. | PR description | N/A (resolved) |
| 2 | `TestOpenMldbSourceIT`, the only test exercising the new client against a real OpenMLDB server, is gated by `@EnabledIfSystemProperty` and never runs in standard CI (confirmed it's also excluded from the default surefire `test` phase by the root `**/*IT.java` exclude). The mock-based unit tests cover dispatch/lifecycle logic well but can't validate the native JNI layer's actual NULL/type behavior end-to-end. | `seatunnel-connectors-v2/connector-openmldb/src/test/java/.../source/TestOpenMldbSourceIT.java` | Low |

Raised by another reviewer: No (no other reviewer has commented on this PR as of this review).

# 5. Merge Recommendation

### Conclusion: Ready to merge

1. Blockers — must be fixed: none found in the source.
2. Recommended fixes — non-blocking:
   - Issue 2: consider, as a future follow-up (not this PR), whether a Testcontainers-based OpenMLDB image could let at least a smoke-level multi-table test run in default CI — understanding this may not be practical given the native SDK's platform constraints.

**CI status at the time of this review:** the previous fork run (`34788427810`) failed exactly one job, `Run / Code style` (`UTClassNameCheckTest`), and the new commit (`1e436ea4208f`) fixes exactly that, via a verified-safe rename that does not change which tests execute under `mvn test` (still excluded by the root `**/*IT.java` surefire exclude either way). The new fork run (`34918561906`) and the corresponding apache-side "Build" check are both still queued/in-progress as of this review — no result to report yet, and per this project's policy I'm not claiming it green until it actually completes.

**Upstream-sync note:** this PR is 5 commits behind `dev` (`46e58fc16`, `bd09f6d9a`, `ae9132fe0`, `41a9c4b8f`, `0d9f9e230`). I checked all five commit messages/paths and none touch `connector-openmldb`, `seatunnel-ci-tools`, or the naming-convention check, so being behind dev is not the cause of (and syncing is not required to fix) the Code style failure that's already addressed above. Syncing is optional housekeeping, not a blocker.

**Overall assessment:** unchanged from my last review — this is a well-engineered PR: the multi-table wiring is validated defensively at config time, native resource management is careful about ownership and concurrent-close safety, several genuine pre-existing correctness bugs are fixed and are now properly disclosed, and the documentation is unusually thorough. The one procedural item from my last review (CI) has a diagnosed root cause with a correct, minimal fix already pushed; once the new CI run completes green, this is good to merge.

**Review by davidzollo — APPROVED** [review]
+1
Good job
